sklearn/preprocessing/_encoders.py

```diff
@@ -421,7 +421,8 @@ def _legacy_fit_transform(self, X):
         dtype = getattr(X, 'dtype', None)
         X = check_array(X, dtype=np.int)
         if np.any(X < 0):
-            raise ValueError("X needs to contain only non-negative integers.")
+            raise ValueError("Instantiate using categories='auto' to"
+                             "handle negative integer categories.")
```

Comment:
May I suggest the following error message that I find more explicit:

OneHotEncoder in legacy mode cannot handle categories encoded as negative integers. Please set categories='auto' explicitly to be able to use arbitrary integer values as category identifiers.

Reply:
Sure, Will do.